File: ascentschooliosapp/ascentschooliosapp/Core/Networking/DTOs/MessagingDTOs.swift
import Foundation

// Phase 92 (Android parity): parent ↔ teacher messaging. Field names/shapes
// mirror the server `MessagingDtos.cs` exactly. Dates arrive as ISO strings
// and are kept as `String?` (formatted for display with `friendlyDate()` /
// a short time formatter), matching the rest of the app's date handling.

// One message in a thread. `senderType` is "parent" | "teacher"; `status` is
// "Active" | "Removed" (removed messages render as "This message was removed.").
struct MessageDto: Decodable, Identifiable {
    let messageId: Int
    let threadId: Int
    let senderType: String
    let senderId: Int
    let senderName: String?
    let body: String
    let status: String?
    let readAt: String?
    let createdAt: String?

    var id: Int { messageId }

    var isRemoved: Bool { (status ?? "").caseInsensitiveCompare("Removed") == .orderedSame }
}

// A parent↔teacher conversation about one child. Used by the teacher inbox.
struct MessageThreadDto: Decodable, Identifiable {
    let threadId: Int
    let studentUniqueId: Int?
    let parentId: Int?
    let studentName: String?
    let admissionNo: String?
    let className: String?
    let sectionName: String?
    let status: String?          // Active | Blocked
    let blockedByType: String?
    let blockedAt: String?
    let lastMessageBody: String?
    let lastMessageAt: String?
    let unreadCount: Int?        // unread for the caller's side
    let createdAt: String?

    var id: Int { threadId }
}

// GET /mobile/teacher/messages/{threadId} — a thread plus its messages.
struct MessageThreadDetailDto: Decodable {
    let thread: MessageThreadDto?
    let messages: [MessageDto]
}

// GET /mobile/messages — everything the parent chat screen opens with.
struct ParentThreadViewDto: Decodable {
    let canMessage: Bool
    let reason: String?            // set when canMessage == false
    let teachers: [String]         // recipient display names
    let threadId: Int?             // nil until the first message
    let status: String?            // Active | Blocked
    let blockedByType: String?     // parent | teacher
    let messages: [MessageDto]
}

// POST send/reply response payload ({ messageId, threadId }).
struct SendMessageResult: Decodable {
    let messageId: Int
    let threadId: Int
}

struct SendMessageRequest: Encodable {
    let body: String
}

struct ReportMessageRequest: Encodable {
    let messageId: Int
    let reason: String?
}

File: API/AscentSchools.Core/DTOs/School/Messaging/MessagingDtos.cs
using System;
using System.Collections.Generic;

namespace AscentSchools.Core.DTOs.School.Messaging
{
    /// <summary>A parent &lt;-&gt; teacher conversation about one child.</summary>
    public class MessageThreadDto
    {
        public int       ThreadId        { get; set; }
        public int       StudentUniqueId { get; set; }
        public int       ParentId        { get; set; }
        public string    StudentName     { get; set; }
        public string    AdmissionNo     { get; set; }
        public string    ClassName       { get; set; }
        public string    SectionName     { get; set; }
        public string    Status          { get; set; }   // Active | Blocked
        public string    BlockedByType   { get; set; }
        public DateTime? BlockedAt       { get; set; }
        public string    LastMessageBody { get; set; }
        public DateTime? LastMessageAt   { get; set; }
        public int       UnreadCount     { get; set; }   // unread for the CALLER's side
        public DateTime  CreatedAt       { get; set; }
    }

    public class MessageDto
    {
        public int       MessageId  { get; set; }
        public int       ThreadId   { get; set; }
        public string    SenderType { get; set; }   // parent | teacher
        public int       SenderId   { get; set; }
        public string    SenderName { get; set; }
        public string    Body       { get; set; }
        public string    Status     { get; set; }   // Active | Removed
        public DateTime? ReadAt     { get; set; }
        public DateTime  CreatedAt  { get; set; }
    }

    /// <summary>A thread plus its messages — what a chat screen opens with.</summary>
    public class MessageThreadDetailDto
    {
        public MessageThreadDto Thread   { get; set; }
        public List<MessageDto> Messages { get; set; }
    }

    /// <summary>
    /// Read-only admin conversation-list row (school web app — Conversations viewer).
    /// Class/section reflect the student's placement in the SELECTED academic year;
    /// TeacherNames are the class teachers assigned for that year.
    /// </summary>
    public class ConversationListItemDto
    {
        public int       ThreadId        { get; set; }
        public int       StudentUniqueId { get; set; }
        public string    StudentName     { get; set; }
        public string    AdmissionNo     { get; set; }
        public string    ClassName       { get; set; }
        public string    SectionName     { get; set; }
        public string    TeacherNames    { get; set; }   // comma-separated assigned class teachers
        public string    Status          { get; set; }   // Active | Blocked
        public string    LastMessageBody { get; set; }
        public DateTime? LastMessageAt   { get; set; }
        public int       MessageCount    { get; set; }
    }

    public class SendMessageRequest
    {
        public string Body { get; set; }
    }

    /// <summary>
    /// A student's current-year placement — everything message routing needs,
    /// in one lookup. Data-layer row shape.
    /// </summary>
    public class StudentClassContextDto
    {
        public int  AcademicYearId  { get; set; }
        public int  ClassId         { get; set; }
        public int? SectionId       { get; set; }
        public int  StudentUniqueId { get; set; }
    }

    public class ReportMessageRequest
    {
        public int    MessageId { get; set; }
        public string Reason    { get; set; }
    }

    /// <summary>Admin review queue row (school web app).</summary>
    public class MessageReportDto
    {
        public int       ReportId       { get; set; }
        public int       MessageId      { get; set; }
        public int       ThreadId       { get; set; }
        public string    ReportedByType { get; set; }
        public int       ReportedById   { get; set; }
        public string    Reason         { get; set; }
        public string    Status         { get; set; }   // Open | Reviewed | Removed
        public string    ReviewedBy     { get; set; }
        public DateTime? ReviewedAt     { get; set; }
        public DateTime  CreatedAt      { get; set; }
        // Context for the reviewer
        public string    MessageBody    { get; set; }
        public string    MessageStatus  { get; set; }
        public string    SenderType     { get; set; }
        public string    SenderName     { get; set; }
        public string    StudentName    { get; set; }
        public string    ClassName      { get; set; }
    }

    public class ResolveReportRequest
    {
        /// <summary>Removed = hide the message from both sides; Reviewed = no action.</summary>
        public string Action { get; set; }   // Removed | Reviewed
    }

    /// <summary>
    /// Everything the parent chat screen needs in one call: whether messaging is
    /// available for this child's class, who it reaches, and the conversation.
    /// </summary>
    public class ParentThreadViewDto
    {
        public bool         CanMessage    { get; set; }
        public string       Reason        { get; set; }   // set when CanMessage = false
        public List<string> Teachers      { get; set; }   // recipient display names
        public int?         ThreadId      { get; set; }   // null until the first message
        public string       Status        { get; set; }   // Active | Blocked
        public string       BlockedByType { get; set; }   // parent | teacher
        public List<MessageDto> Messages  { get; set; }
    }
}
